pkg/server/api/debug/v1/service.go

```diff
@@ -0,0 +1,172 @@
+package debug
+
+import (
+	"context"
+	"crypto/x509"
+	"sync"
+	"time"
+
+	"github.com/spiffe/go-spiffe/v2/bundle/x509bundle"
+	"github.com/spiffe/go-spiffe/v2/spiffeid"
+	"github.com/spiffe/go-spiffe/v2/svid/x509svid"
+	"github.com/spiffe/spire/pkg/server/api/rpccontext"
+	"github.com/spiffe/spire/pkg/server/plugin/datastore"
+	"github.com/spiffe/spire/pkg/server/svid"
+	"github.com/spiffe/spire/proto/spire/api/server/debug/v1"
+	debug_pb "github.com/spiffe/spire/proto/spire/api/server/debug/v1"
+	"github.com/spiffe/spire/proto/spire/types"
+	"github.com/spiffe/spire/test/clock"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+const (
+	cacheExpiry = 5 * time.Second
+)
+
+// RegisterService registers debug service on provided server
+func RegisterService(s *grpc.Server, service *Service) {
+	debug_pb.RegisterDebugServer(s, service)
+}
+
+// Config configurations for debug service
+type Config struct {
+	Clock        clock.Clock
+	DataStore    datastore.DataStore
+	SVIDObserver svid.Observer
+	TrustDomain  spiffeid.TrustDomain
+	Uptime       time.Time
+}
+
+// New creates a new debug service
+func New(config Config) *Service {
+	return &Service{
+		clock:  config.Clock,
+		ds:     config.DataStore,
+		so:     config.SVIDObserver,
+		td:     config.TrustDomain,
+		uptime: config.Uptime,
+
+		getInfoResp: &infoResp{},
+	}
+}
+
+// Service implements debug server
+type Service struct {
+	clock  clock.Clock
+	ds     datastore.DataStore
+	so     svid.Observer
+	td     spiffeid.TrustDomain
+	uptime time.Time
+
+	getInfoResp *infoResp
+}
+
+type infoResp struct {
+	mtx  sync.Mutex
+	resp *debug_pb.GetInfoResponse
+	ts   time.Time
+}
+
+// GetInfo gets SPIRE Server debug information
+func (s *Service) GetInfo(ctx context.Context, req *debug_pb.GetInfoRequest) (*debug_pb.GetInfoResponse, error) {
+	log := rpccontext.Logger(ctx)
+
+	s.getInfoResp.mtx.Lock()
+	defer s.getInfoResp.mtx.Unlock()
+
+	// Update cache when expired or does not exists
+	if s.getInfoResp.ts.IsZero() || s.clock.Now().Sub(s.getInfoResp.ts) >= cacheExpiry {
+		nodes, err := s.ds.CountAttestedNodes(ctx, &datastore.CountAttestedNodesRequest{})
+		if err != nil {
+			log.WithError(err).Error("Failed to count agents")
+			return nil, status.Errorf(codes.Internal, "failed to count agents: %v", err)
```

Comment:
We could use the `MakeErr()` method to format, log and return the error in this and all the following early returns.

```suggestion
			return nil, api.MakeErr(log, codes.Internal, "failed to count agents", err)
```